### PR TITLE
1bit for Issue #13

### LINKED ISSUE #13: Train 1bit baseline
## Description

1. Train the 1 bit baseline with naive 1 bit quantisation (i.e. change BitDistiller int-asym from 2 to 1)
2. Upload the model/logs to hugging face with a sensible informative name (see #12 for of what to possibly include). Save a screenshot of the tensorboard plots.
3. Run the faster evals like PPL and language benchmarks that aren't MMLU/hellaswag, check the performance against the baseline.
4. Put the plot and metrics on the model card or in a sensible file.

## Checklist before resolving issue
- [ ] I have included the issue number in my PR description
- [ ] I have rebased/pulled in changes from `main` and fixed any conflicts
- [ ] If I have changed the models/training code, I have run `dry_run.sh` on a CUDA device successfully

### PR DESCRIPTION
**Pull request by fredericowieser**
Add the implementation for a 1bit {0, 1} version of the asymmetric clipping described using 'int' format in the original BitDistiller paper

Closes #13

### CODE REVIEW

File: train/train.py [hunk 1]
```diff
@@ -337,20 +356,24 @@ def train():
             device_map=device_map,
             max_memory=max_memory,
         )
-        teacher_model.eval()
+        teacher_model.eval()  # Set teacher model to evaluation mode
```
Comment (BrownianNotion) [inline train/train.py:359]: Unnecessary comment

File: quantization/quantizer.py [hunk 2]
```diff
@@ -180,6 +218,57 @@ def forward(self, x):
...
+            assert org_w_shape[-1] % self.q_group_size == 0 # Last dimension must be divisible by q_group_size
+            x = x.reshape(-1, self.q_group_size)
+        assert x.dim() == 2 # Tensor should be 2D after grouping
+
+        # Compute the per-row min and max
+        max_val = x.amax(dim=1, keepdim=True)  # (num_groups, 1)
+        min_val = x.amin(dim=1, keepdim=True)  # (num_groups, 1)
+
+        # Define the min and max integer values
+        # TODO: Move This Hardcoded Value somewhere so it's not
```
Comment (BrownianNotion) [inline quantization/quantizer.py:242]: Can move this into init method and store

File: quantization/quantizer.py [hunk 3]
```diff
@@ -180,6 +218,57 @@ def forward(self, x):
...
+        # Transform float -> integer in {0,1}
+        # using Round STE, clamp to [0,1]
+        x_int = torch.clamp(
+            Round.apply(x / scale) + zeros, min_int, max_int
+        )
+
+        # Transform integer -> float by removing zero_point, then scaling
+        x_q = (x_int - zeros) * scale
+
+        # Double-check no NaNs introduced
```
Comment (BrownianNotion) [inline quantization/quantizer.py:265]: Again, a few unnecessary comments here (and below for reshape/assert statement on nans, train file seed/parse comments etc.)

Commenting is definitely helpful for explaining the code flow, particularly things that aren't obvious from the way the code is written (eg. mathematical motivation), but I think this type of comment is just clutter, and I think it'd be good to remove a lot of these for smaller diff.

File: quantization/quantizer.py [hunk 4]
```diff
@@ -68,6 +68,44 @@ def pseudo_quantize_tensor(w, n_bit=8,
     else:
         return w
 
+def pseudo_quantize_tensor_1bit(w,
+                           zero_point=True, q_group_size=-1,
+                           inplace=False,
+                           get_scale_zp=False
+                           ):
+    n_bit = 1
```
Comment (BrownianNotion) [inline quantization/quantizer.py:76]: delete unused variable

File: quantization/quantizer.py [hunk 5]
```diff
@@ -68,6 +68,44 @@ def pseudo_quantize_tensor(w, n_bit=8,
     else:
         return w
 
+def pseudo_quantize_tensor_1bit(w,
```
Comment (BrownianNotion) [inline quantization/quantizer.py:71]: Probably would have been easier to reviewer/cleaner in code to just add an if clause for the original function and inherit the original SteInt2AsymQuantizer`, changing where needed for 1 bit. If we end up changing these inner workings a lot, that's fine, but perhaps in future, let's aim for minimal changes until we're sure we need them.

**Review by BrownianNotion — APPROVED** [review]
Ideally delete the unneeded comments

**Comment by BrownianNotion** [comment]
Also, specify issue https://github.com/BrownianNotion/BitDistiller/issues/13 in PR title

**Comment by BrownianNotion** [comment]
Note for next time: Add clipped weights to the repo for saving time + reproducibility. Will include section in readme about this.

File: quantization/autoclip.py [hunk 6]
```diff
@@ -5,7 +5,11 @@
 import os
 import sys
 from clip_utils import *
-from quantizer import pseudo_quantize_tensor, pseudo_quantize_n2f3_tensor
+from quantizer import (
```
[the remark this answers is not in the document]
Comment (fredericowieser) [reply quantization/autoclip.py:8]: black pre commit hook?

File: quantization/quantizer.py (hunk 5 above)
Comment (fredericowieser) [reply quantization/quantizer.py:71]: I agree with the statement you are making, this was probably overkill, but I wanted to reduce any unnecessary floating point errors by removing these unnecessary powers and divisions made in the more general code...

General Code:
```python
    if zero_point:
        max_val = w.amax(dim=1, keepdim=True)
        min_val = w.amin(dim=1, keepdim=True)
        max_int = 2 ** n_bit - 1
        min_int = 0
        scales = (max_val - min_val).clamp(min=1e-5) / max_int
        zeros = (-torch.round(min_val / scales)).clamp_(min_int, max_int)
```

1 Bit:
```python
    if zero_point:
        max_val = w.amax(dim=1, keepdim=True)
        min_val = w.amin(dim=1, keepdim=True)
        max_int = 1
        min_int = 0
        scales = (max_val - min_val).clamp(min=1e-5)
        zeros = (-torch.round(min_val / scales)).clamp_(min_int, max_int)
```

I suppose this could be done in an if statement in the original function instead to be fair though, could raise this as tech debt or if you feel it's important can open this PR again and put it in. Happy to do either :)
